### DISCUSSION TITLE
[mini.align] LibreOffice Calc tables to Markdown

### DISCUSSION THREAD

**Discussion by raffaelemancuso** [nvim-mini/mini.nvim#1047] [Q&A]
I have made this keybinding that converts a table copied from LibreOffice Calc into a pandoc markdown table:

This:

* Escape `%` characters
* Convert tabs into `|` (column separators)
* Put a `|` at the end of each line
* Put a `|` at the start of each line
* Put the cursor at the first line
* Copy the first line and paste it right after
* Substitute the text in the second line with `---` (define the first line as column headers)
* Select the entire table
* MISSING: Aligns the `|` on top of each other

```
vim.api.nvim_set_keymap(
  "n",
  "<LocalLeader>t",
  "vip:s/%/\\%/ge<CR>vip:s/\t/|/g<CR>vip:s/$/|/<CR>vip:s/^/|/<CR>vipo<Esc>yyp:s/.\\{-}|/---|/g<CR>dt|:nohlsearch<CR>vip",
  {
    noremap = true,
    desc = "Table: excel2md"
  }
)
```

The only thing I need is to align the `|` (column separators) on top of each other.

How can I do that programmatically with mini.align, by calling a LUA function?

**Comment by echasnovski** [comment]
Might appending something like `ga|vip` work? Here `ga|` aligns restored paragraph visual selection and `vip` again restores visual selection. This would require using `noremap = false` though.

If you *have to* use Lua function, take a look at [`MiniAlign.align_strings()`](https://github.com/echasnovski/mini.nvim/blob/861c8aa3c68063057707f874e3ae09bf313fd299/doc/mini-align.txt#L551). But using it in this use case (single mapping) doesn't look like straightforward.

With example of "before" and "after" the answer might have been more useful.

**Reply by raffaelemancuso** [reply]
Sorry for the lack of an example.

Here is an example before:

```
	2020	2021
Revenues	100	120
Expenses	80	90
Margin	20	30
Margin (%)	20.00%	25.00%
```

Here is  the same  example after:

```
|            | 2020   | 2021   |
| ---        | ---    | ---    |
| Revenues   | 100    | 120    |
| Expenses   | 80     | 90     |
| Margin     | 20     | 30     |
| Margin (%) | 20.00% | 25.00% |
```

> This would require using noremap = false though.

Yeah the whole point is  that I cannot use `noremap = false`.

I ended up with this final mapping which uses [godlygeek's tabular](https://github.com/godlygeek/tabular).

It would be nice if mini.align would have easy commands like this.

```
vim.api.nvim_set_keymap(
  "n",
  "<LocalLeader>t",
  '"+pvip:s/%/\\%/ge<CR>vip:s/\t/|/g<CR>vip:s/$/|/<CR>vip:s/^/|/<CR>vipo<Esc>yyp:s/.\\{-}|/---|/g<CR>dt|:nohlsearch<CR>vip:Tabularize /|<CR>vip<',
  {
    noremap = true,
    desc = "Paste LOCalc 2 MD"
  }
)
```

**Reply by echasnovski** [reply]
> Yeah the whole point is that I cannot use `noremap = false`.

Is there a particular reason for it? I'd say that the whole mapping uses uncontrovercial keys.

**Reply by echasnovski** [reply]
Here is how I would do it with Lua:

```lua
local excel_to_md = function()
  -- Initiate marks
  vim.cmd('normal! vip\27')

  -- Transform body
  vim.cmd([['<,'>s/%/\%/ge]])
  vim.cmd([['<,'>s/\t/|/g]])
  vim.cmd([['<,'>s/$/|/]])
  vim.cmd([['<,'>s/^/|/]])

  -- Transform header
  vim.cmd('normal! vipo\27yyp')
  vim.cmd([[s/.\{-}|/---|/g]])
  vim.cmd('normal! dt|')

  -- Align (requires 'mini.align')
  local from, to = vim.fn.line("'<"), vim.fn.line("'>")
  local lines = vim.api.nvim_buf_get_lines(0, from - 1, to, false)
  local aligned_lines = require('mini.align').align_strings(lines, { split_pattern = '|', merge_delimiter = ' ' })
  vim.api.nvim_buf_set_lines(0, from - 1, to, false, aligned_lines)

  -- Cleanup and restore visual selection
  vim.cmd('nohlsearch')
  vim.cmd('normal! vip')
end

vim.keymap.set('n', '<Leader>yl', excel_to_md, { desc = 'Table: excel2md' })
```

Although it takes much more space, I'd argue that this is at the same time much more readable and understandable.

**Reply by raffaelemancuso** [reply]
Thank you very much.

I'm also exploring other options for this like [markdown-table-mode](https://github.com/Kicamon/markdown-table-mode.nvim)
